text/014-get-page-recuction.md

```diff
@@ -0,0 +1,52 @@
+# Reduction of the GetPage@LSN problem to GetPage@LatestLSN
+
+## 1. Motivation
+
+The Socrates paper [1] from Microsoft mentions that "Serving GetPage@LSN requests is also straightforward. The Page Server simply follows the protocol of Section 4.4", and I wish I could say the same about our system. Our storage format is complicated and still unfinished. At first glance it looks like our problem is significantly harder because our GetPage@LSN **cannot** return a newer version than what was requested. We're serving historical data, and Socrates is serving latest data.
+
+My goal here is to show that our GetPage@LSN problem reduces to GetPage@LatestLSN.
+
+## 2. Context
+Why does our pageserver need to serve historical data? Because we want the ability to serve read nodes at some point, and they will lag. Not all read replicas will reuse an existing pageserver, but there are at least a few cases where that seems useful:
+1. Not much read traffic, but read replica helps with isolating read-only queries
+2. Lots of read traffic, and primary is already the biggest instance available
+
+These two use cases are enough to justify our desire to keep our options open.
+
+## 3. Definitions
+The following terms are used in section 4.
+
+Definition: Let `hot_wal_size` be a very conservative upper bound on how much our read replicas can lag behind, in terms of WAL size. One way to derive it is `hot_wal_size = max_lag_seconds * safekeeper_ssd_troughput`. Example: if we (hypothetically) want to allow 10 seconds of lag, and safekeepers nodes are writing 1 GB/s on 8 SSDs, then `hot_wal_size = 80 GB`.
+
+Definition: A page version is **hot** if it has been modified within the last `hot_wal_size` gigabytes of wal, and **cold** otherwise.
+
+## 4. The problem reduction
+We define 2 helper lemmas and use them in the final proof below.
+
+### 4.1 Lemma 1
+Statement: The pageserver will only get two kinds of get_page queries:
+a) get a hot page at some LSN within `hot_wal_size`
+b) get the latest cold version of a page
+
+Proof: It follows from the definition of `hot_wal_size` as an upper bound on replica lag. 
+
+### 4.2 Lemma 2
+Statement: Answering queries of type (a) is easy.
+
+Proof: The `hot_wal_size` fits in RAM. One implementation of (a) is to just read the `PageReconstructData` from ram, and optimize our wal redo enough so that get_page is pretty fast. Some people believe wal redo can be as fast as memcopy. Others believe that there alternative in-memory data structures that obviate the need for redo. I'm sure there are other approaches, but in the end this is easier than finding `PageReconstructData` from disk. I refrain from picking a solution in order to avoid strawman arguments.
```

Comment:
I do no agree with this lemme. Let wal records which has to be applied fit in RAM. But them has to be applied to some page image. And this page may not be present in RAM. So we have to read if from disk. And it can be quite expensive.

Reply:
Read the page image from disk, read the wal from ram. Reading 8kb from disk is way faster than reading scattered megabytes of scattered PageReconstructData from disk.

Reply:
I also don't agree with this lemma. The definitions used in (3) apply to safekeepers; not necessarily PageServers. PageServers need to index the received WAL and thus might need potentially significant overhead for this index. Therefore, the size of the WAL at the safekeeper is not equivalent to the size of the WAL at the PageServer.

Furthermore; that something fits in RAM doesn't mean that it is easily accessed. Finding an arbitrary value in the digits of Pi is not somehow trivializable, even if all digits of Pi were somehow pre-loaded into RAM: the operation still is O(n); and O(n) on the example's 80GB of data is not something to trivialize with "it is easy".

Sure, linear search might be easy to implement, but it is not performant, and we care a lot about performance.